File: D3/4047.cpp
// 제목 : 영준이의 카드 카운팅
// 링크 : https://swexpertacademy.com/main/code/problem/problemDetail.do?contestProbId=AWIsY84KEPMDFAWN
// 메모리 :  12,684 kb
// 실행시간 : 6 ms

#define _CRT_SECURE_NO_WARNINGS 
#include<iostream>
#include<string>

using namespace std;

int main(int argc, char** argv)
{
    int test_case;
    int T;  //테스트 케이스 수

    cin >> T;

    for (test_case = 1; test_case <= T; ++test_case)
    {
        //카드를 저장할 배열
        int** cards = new int* [4];
        for (int i = 0; i < 4; i++) {
            cards[i] = new int[14];
            fill(cards[i], cards[i] + 14, 0);
        }

        //카드 정보 입력받기
        string s;
        cin >> s;
        int pos = 0;
        bool chk = true;
        //3글자씩 끊어서 카드 정보 해석
        while (pos < s.length()) {
            //첫 글자로 카드 무늬 읽어서
            if (s[pos] == 'S') {
                //그 무늬 카드 갯수를 하나 늘리고
                cards[0][0]++;
                //그 무늬 그 숫자 카드 갯수를 하나 늘리고
                cards[0][stoi(s.substr(pos + 1, 2))]++;
                //같은 무늬 같은 숫자 카드가 둘 이상이라면 그만 읽기
                if (cards[0][stoi(s.substr(pos + 1, 2))] > 1) {
                    chk = false;
                    break;
                }
            }
            else if (s[pos] == 'D') {
                cards[1][0]++;
                cards[1][stoi(s.substr(pos + 1, 2))]++;
                if (cards[1][stoi(s.substr(pos + 1, 2))] > 1) {
                    chk = false;
                    break;
                }
            }
            else if (s[pos] == 'H') {
                cards[2][0]++;
                cards[2][stoi(s.substr(pos + 1, 2))]++;
                if (cards[2][stoi(s.substr(pos + 1, 2))] > 1) {
                    chk = false;
                    break;
                }
            }
            else if (s[pos] == 'C') {
                cards[3][0]++;
                cards[3][stoi(s.substr(pos + 1, 2))]++;
                if (cards[3][stoi(s.substr(pos + 1, 2))] > 1) {
                    chk = false;
                    break;
                }
            }
            pos += 3;
        }
        //중복카드가 없다면 남은 카드 갯수 출력, 있다면 오류메시지 출력
        if (chk) cout << '#' << test_case << ' ' << 13 - cards[0][0] << ' ' << 13 - cards[1][0] << ' ' << 13 - cards[2][0] << ' ' << 13 - cards[3][0] << '\n';
        else cout << '#' << test_case << ' ' << "ERROR" << '\n';

        for (int i = 0; i < 4; i++) delete[] cards[i];
        delete[] cards;
    }
    return 0;
}

File: D3/4047.java
// 제목 : 영준이의 카드 카운팅
// 링크 : https://swexpertacademy.com/main/code/problem/problemDetail.do?contestProbId=AWIsY84KEPMDFAWN
// 메모리 :  18,408 kb
// 실행시간 : 104 ms

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;
  
class Solution
{
	
	public static void main(String args[]) throws IOException
    {
		//빠른 입출력을 위한 BufferedReader br과 StringBuilder answerSheet
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        
        //테스트 케이스 수
        int T = Integer.parseInt(br.readLine());
        
        //각각의 테스트 케이스 마다
        for(int test_case = 1; test_case <= T; test_case++)
        {
        	//카드 정보 입력받기
        	String input = br.readLine();
        	
        	boolean[][] duplicated = new boolean[4][14];	//무늬와 숫자별로 중복확인할 배열
        	int[] required = {13,13,13,13};					//무늬별 필요한 카드 수
        	boolean possible = true;						//중복이 있었는지 없었는지
        	
        	//세 글자씩 카드 정보 읽어오기
        	for(int i=0;i<input.length();i+=3) {
        		//무늬 읽기
        		int type = 3;
        		switch(input.charAt(i)) {
        		case 'S': type=0; break;
        		case 'D': type=1; break;
        		case 'H': type=2; break;
        		}
        		//숫자 읽기
        		int num = Integer.parseInt(input.substring(i+1, i+3));
        		
        		//중복이라면, 중복이라고 표시하고 그만두기
        		if(duplicated[type][num]) {
        			possible = false;
        			break;
        		}
        		
        		//중복이 아니라면 카드가 있다고 표시하고, 남은 카드 숫자 줄이기
        		duplicated[type][num] = true;
        		required[type]--;
        	}
        	
        	//출력문 저장
        	sb.append('#').append(test_case).append(' ');
        	if(possible) {
        		sb.append(required[0]).append(' ').append(required[1]).append(' ').append(required[2]).append(' ').append(required[3]).append('\n');
        	}
        	else {
        		sb.append("ERROR\n");
        	}
        	

        }
        //출력
        System.out.printf(sb.toString());
    }
}